lib/ukrust/Makefile.uk

```diff
@@ -0,0 +1,64 @@
+$(eval $(call addlib_s,libukrust,$(CONFIG_LIBUKRUST)))
+
+CINCLUDES-$(CONFIG_LIBUKRUST) += -I$(LIBUKRUST_BASE)/ukrust_sys
+
+RUSTCFLAGS-y += --extern ukrust_sys -L $(LIBUKRUST_BUILD)/ --extern alloc \
+		--target $(LIBUKRUST_BASE)/target.json \
+		-Zallow-features=allocator_api,bench_black_box,concat_idents,global_asm,try_reserve
+
+
+export RUST_BINDINGS_FILE=$(LIBUKRUST_BUILD)/bindings_generated.rs
+RUSTC_SYSROOT = $(shell $(RUSTC)  --print sysroot)
+RUST_LIB_SRC ?= $(RUSTC_SYSROOT)/lib/rustlib/src/rust/library
+
+# Common flags
+RUST_UK_FLAGS = --target=$(LIBUKRUST_BASE)/target.json --emit=dep-info,obj,metadata --edition=2018 \
+	-Cpanic=abort -Cembed-bitcode=n -Clto=n -Crpath=n \
+	-Cforce-unwind-tables=n -Ccodegen-units=1 \
+	-Zbinary_dep_depinfo=y -Zsymbol-mangling-version=v0 \
+	-Dunsafe_op_in_unsafe_fn -Drust_2018_idioms \
+	--crate-type rlib
+
+# Generate rust bindings for the Unikraft API
+$(LIBUKRUST_BUILD)/bindings_generated.rs: $(LIBUKRUST_BASE)/ukrust_sys/bindings_helper.h
+	$(call build_cmd, GEN, libukrust, $(notdir $@), \
+		bindgen -o $(LIBUKRUST_BUILD)/bindings_generated.rs --use-core --ctypes-prefix crate::c_types \
+		--with-derive-default --no-debug '.*' --size_t-is-usize \
+		$(LIBUKRUST_BASE)/ukrust_sys/bindings_helper.h -- $(CINCLUDES) \
+		$(CINCLUDES-y) -nobuiltininc -DMODULE)
+
+# Build the ukrust_sys crate, contains the bindings and other useful functions
+$(LIBUKRUST_BUILD)/ukrust_sys.o: $(LIBUKRUST_BASE)/ukrust_sys/lib.rs $(LIBUKRUST_BUILD)/core.o $(LIBUKRUST_BUILD)/compiler_builtins.o
+	$(call build_cmd, BUILD, libukrust, $(notdir $@), \
+	$(RUSTC) $(RUST_UK_FLAGS) --extern alloc  \
+		--out-dir $(LIBUKRUST_BUILD) -L $(LIBUKRUST_BUILD) -A dead_code \
+		-A unused_doc_comments --crate-name ukrust_sys $(LIBUKRUST_BASE)/ukrust_sys/lib.rs)
+
+
+$(LIBUKRUST_BUILD)/alloc.o: $(RUST_LIB_SRC)/alloc/src/lib.rs
+	$(call build_cmd, BUILD, libukrust, $(notdir $@), \
+	$(RUSTC) $(RUST_UK_FLAGS)  \
+		--out-dir $(LIBUKRUST_BUILD) -L $(LIBUKRUST_BUILD) -A dead_code \
+		-A unused_doc_comments --crate-name alloc $(RUST_LIB_SRC)/alloc/src/lib.rs)
+
+$(LIBUKRUST_BUILD)/core.o: $(RUST_LIB_SRC)/alloc/src/lib.rs
+	$(call build_cmd, BUILD, libukrust, $(notdir $@), \
+	$(RUSTC) $(RUST_UK_FLAGS) \
+		--out-dir $(LIBUKRUST_BUILD) -L $(LIBUKRUST_BUILD) -A dead_code \
+		-A unused_doc_comments --crate-name core $(RUST_LIB_SRC)/core/src/lib.rs)
+
+$(LIBUKRUST_BUILD)/compiler_builtins.o: $(LIBUKRUST_BASE)/compiler_builtins.rs $(LIBUKRUST_BUILD)/core.o
+	$(call build_cmd, BUILD, libukrust, $(notdir $@), \
+	$(RUSTC) $(RUST_UK_FLAGS) \
+		--out-dir $(LIBUKRUST_BUILD) -L $(LIBUKRUST_BUILD) -A dead_code \
+		-A unused_doc_comments --crate-name compiler_builtins $(LIBUKRUST_BASE)/compiler_builtins.rs)
+
+UK_PREPARE-$(CONFIG_LIBUKRUST) += $(LIBUKRUST_BUILD)/bindings_generated.rs $(LIBUKRUST_BUILD)/ukrust_sys.o \
+		$(LIBUKRUST_BUILD)/alloc.o $(LIBUKRUST_BUILD)/core.o $(LIBUKRUST_BUILD)/compiler_builtins.o
```

Comment:
Do these objects need to be compiled before compiling anything else in rust? If so, can you put a short comment explaining this. Otherwise, we could register the object files to `UK_OLIBS-$(CONFIG_LIBUKRUST)`

Reply:
Yes, anything else uses the symbols defined in these objects. They're standalone since they depend on no crate.